In-app notification API

e32wong opened the pull request:
Changes:
- Added support for notification API for two types of actions:
  - New opinions
  - Opinion agree/disagree
- Implemented a new pull-down-to-refresh composable for the home feed and notification page
  - Changed the loading animation to use `q-inner-loading` instead since we no longer use Quasar's pull-down-to-refresh component due to it being buggy
- Moved the new conversation page's add poll button to the top menu bar as a temporary iOS fix (we will fix it by applying a layout system that is similar to the notification page later)

nicobao commented:
>   * Changed the loading animation to use `q-inner-loading` instead since we no longer use Quasar's pull-down-to-refresh component due to it being buggy
Interesting, can you tell me what was buggy in Quasar's pull to refresh component?

nicobao commented:
![image](https://github.com/user-attachments/assets/ed9725bd-b4ff-41fa-9666-a560d2b31d62)

The profile list of conversation is broken for some reasons. I don't know if it's related to the PR or not.

the backend does return the data, but the frontend does not display it:
![image](https://github.com/user-attachments/assets/0d6344b0-af00-4c9d-a194-f9cc0ba56624)

e32wong commented:
> > * Changed the loading animation to use `q-inner-loading` instead since we no longer use Quasar's pull-down-to-refresh component due to it being buggy
> 
> Interesting, can you develop about what was buggy in Quasar's pull to refresh component?

The problem started when we switched to a more standard infinite scrolling system for the home feed and notification page using VueUse's `useInfiniteScroll`. Infinite scrolling had been problematic because when we are using Quasar's q-layout, the page layout system has no predefined height, so it had to be changed to a fixed height div to ensure that user scrolls are detected properly. There were two options. One is to use Quasar's own infinite scrolling system, the other is to roll with `useInfiniteScroll`. We used to use Quasar's infinite scrolling component, don't remember why but we dropped it due to some issues. So now in this PR we are rolling with `useInfiniteScroll` because our own infinite scroll implementation for the front page was too complicated.

So I switched up the two pages to mount the infinite scrolling system to a full screen fixed height div. When Quasar's pull down to refresh component is mounted on a fixed height div, it caused the infinite scrolling to be laggy/jumpy, and the pull down to refresh area is also not configurable and the selection area seems to be too large when it is mounted on a div container.

e32wong commented:
> ![image](https://private-user-images.githubusercontent.com/46368291/407263310-ed9725bd-b4ff-41fa-9666-a560d2b31d62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNjE1MTAsIm5iZiI6MTczODA2MTIxMCwicGF0aCI6Ii80NjM2ODI5MS80MDcyNjMzMTAtZWQ5NzI1YmQtYjRmZi00MWZhLTk2NjYtYTU2MGQyYjMxZDYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI4VDEwNDY1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyYzM4M2M1ZTBiZGFiZjE0NmY4NWRkMTJhMzc3NjE2ZTJmYjA0Zjc0NjNmMGQwMDZiZjc1MjM5NTQwZDVlNDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DKBvjqM3IaMkKo_3vs9s8Yr-M-um12ndviBwXE0NvMo)
> 
> The profile list of conversation is broken for some reasons. I don't know if it's related to the PR or not.
> 
> the backend does return the data, but the frontend does not display it: ![image](https://private-user-images.githubusercontent.com/46368291/407263871-0d6344b0-af00-4c9d-a194-f9cc0ba56624.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNjE1MTAsIm5iZiI6MTczODA2MTIxMCwicGF0aCI6Ii80NjM2ODI5MS80MDcyNjM4NzEtMGQ2MzQ0YjAtYWYwMC00YzlkLWExOTQtZjljYzBiYTU2NjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI4VDEwNDY1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkNjMyNzI5ZTA0Y2U4MmIyOTU3YzVjMDRiNDI1NTllZGJmMTdjOTc4NmIxMGM4OTg1MTRmZjcwNDJmODUwYWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DbKL49Bdmlh9P3aaVihfVBZVOhlvxNab4kAE61GbZiI)

This is now fixed. It was displaying the wrong data.

nicobao commented:
> > > * Changed the loading animation to use `q-inner-loading` instead since we no longer use Quasar's pull-down-to-refresh component due to it being buggy
> > 
> > 
> > Interesting, can you develop about what was buggy in Quasar's pull to refresh component?
> 
> The problem started when we switched to a more standard infinite scrolling system for the home feed and notification page using VueUse's `useInfiniteScroll`. Infinite scrolling had been problematic because when we are using Quasar's q-layout, the page layout system has no predefined height, so it had to be changed to a fixed height div to ensure that user scrolls are detected properly. There were two options. One is to use Quasar's own infinite scrolling system, the other is to roll with `useInfiniteScroll`. We used to use Quasar's infinite scrolling component, don't remember why but we dropped it due to some issues. So now in this PR we are rolling with `useInfiniteScroll` because our own infinite scroll implementation for the front page was too complicated.
> 
> So I switched up the two pages to mount the infinite scrolling system to a full screen fixed height div. When Quasar's pull down to refresh component is mounted on a fixed height div, it caused the infinite scrolling to be laggy/jumpy, and the pull down to refresh area is also not configurable and the selection area seems to be too large when it is mounted on a div container.

Makes sense. _ wonder what was the reason why we didn't use the infinite scrolling from quasar. It would be interesting _at some point_ to fork the Quasar `PullToRefresh` component source code, and make our own custom pull to refresh component by fixing it. Because otherwise overall the look and feel was great. There are unfortunately no solid pull to refresh js library that is actively maintained.

nicobao commented:
- Change `<username> have disagree to your opinion` to `<username> have disagreeD to your opinion`  
- Change `<username> have agree to your opinion` to `<username> have agreeD to your opinion`

nicobao commented:
![image](https://github.com/user-attachments/assets/10ef4c68-f85a-4e58-ab27-b87ab092c2bc)

Two same conversation in the feed! with the same url (OK it happened after I deleted the back multiple times, so it may due to a weird sync error though, not priority)

nicobao commented:
![image](https://github.com/user-attachments/assets/dd424e40-c42b-4c52-b185-7133df22d7bb)


keep the inside text area white

nicobao commented:
this should not have a grey background:

![image](https://github.com/user-attachments/assets/fa849838-9095-402f-b89a-790f53e63592)


But a white one, like here:

![image](https://github.com/user-attachments/assets/03421cb1-99b6-446c-a378-bebdeca87f8f)

e32wong commented:
> ![image](https://private-user-images.githubusercontent.com/46368291/407461909-dd424e40-c42b-4c52-b185-7133df22d7bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwOTI0MDUsIm5iZiI6MTczODA5MjEwNSwicGF0aCI6Ii80NjM2ODI5MS80MDc0NjE5MDktZGQ0MjRlNDAtYzQyYi00YzUyLWIxODUtNzEzM2RmMjJkN2JiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI4VDE5MjE0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4MzM5NTFiYTliYTkwMWNjODQxMjFhYzI1OWZjYTdmYjNiYmEzZjczYmY5MzRhMjY3OTA2ODQxNWIyZjEwMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6xY1aOIK-zZlvRcMNAluePC5j9uYSJ0egRLHB4XGBM)
> 
> keep the inside text area white (including the background of the font modifiers)

Fixed by adding a white background.

e32wong commented:
> this should not have a grey background:
> 
> ![image](https://private-user-images.githubusercontent.com/46368291/407462815-fa849838-9095-402f-b89a-790f53e63592.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwOTIzNjAsIm5iZiI6MTczODA5MjA2MCwicGF0aCI6Ii80NjM2ODI5MS80MDc0NjI4MTUtZmE4NDk4MzgtOTA5NS00MDJmLWI4OWEtNzkwZjUzZTYzNTkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI4VDE5MjEwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlN2Q4NTUzNWM0ZTliZjQ2NGVjMWFjZDZhYTdmZjdjOGU0ZjRlNmM0YzYxYjQ1MTczYWI3Njc2NmJlMDc5MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MDBJ5z-Apa7-B4rifmenu5cURBUMETDsDW2hP2uiqmM)
> 
> But a white one, like here:
> 
> ![image](https://private-user-images.githubusercontent.com/46368291/407462983-03421cb1-99b6-446c-a378-bebdeca87f8f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwOTIzNjAsIm5iZiI6MTczODA5MjA2MCwicGF0aCI6Ii80NjM2ODI5MS80MDc0NjI5ODMtMDM0MjFjYjEtOTliNi00NDZjLWEzNzgtYmViZGVjYTg3ZjhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI4VDE5MjEwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmNWU0YTIyMzZhNjU5NzViNTIwMzEzMjBiNzYzZTI0NjZiNDljNzk3MmE2Zjc4YWM2MWNiYmExMTUyY2EzN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gFb4Kuk76_U_J7dm83ZNYgBLMl7UJW-tgHYu_nUSzfs)

Fixed with a white background.

nicobao reviewed (commented):
I would like us to create a prettier `pull-to-refresh` component probably only after the stack of UI/UX change you're already doing, and after we work to release to the store

nicobao commented:
For info the rationale for renaming is:
- user as simple names as possible, `user` seemed redundant here, as I don't expect we'll have another type of notification (if we built-in monitoring to admins/mods, we can name it differently)
- I suggest to use "fetch" for endpoints/functions that are used to fetch possibly large amount of data (infinite scrolling etc), while we use "get" for more static data.

e32wong commented on services/api/src/schema.ts:
```diff
@@ -1385,3 +1403,85 @@ export const opinionModerationTable = pgTable("opinion_moderation", {
+    },
+    (t) => {
+        return {
+            userIdx: index("user_idx_notification").on(t.userId),
```
No PostgreSQL doesn't create indexes automatically for foreign keys.

nicobao commented:
Also for naming I suggest we use `authorId` when we want to say that the userId refers to the.. author of said content data (specifically for opinion/vote/conversation, meaning any data actually created by the user), while we use "userId" for anything else (i.e: notification generated for a specific user, static phone/passport data associated with a specific user...etc)
